consensus/state_processing/src/per_epoch_processing/single_pass.rs

```diff
@@ -0,0 +1,628 @@
+                &mut next_epoch_activation_queue,
+                state_ctxt,
+                spec,
+            )?;
+        }
+
+        // `process_slashings`
+        if conf.slashings {
+            process_single_slashing(balance, validator, slashings_ctxt, state_ctxt, spec)?;
+        }
+
+        // `process_effective_balance_updates`
+        if conf.effective_balance_updates {
+            process_single_effective_balance_update(
+                *balance,
+                validator,
+                validator_info,
+                &mut next_epoch_total_active_balance,
+                &mut next_epoch_cache,
+                progressive_balances,
+                effective_balances_ctxt,
+                state_ctxt,
+                spec,
+            )?;
+        }
+    }
+
+    if conf.effective_balance_updates {
+        state.set_total_active_balance(next_epoch, next_epoch_total_active_balance);
+        *state.epoch_cache_mut() = next_epoch_cache.into_epoch_cache(
+            next_epoch_total_active_balance,
+            next_epoch_activation_queue,
+            spec,
+        )?;
+    }
+
+    Ok(summary)
+}
+
+fn process_single_inactivity_update(
+    inactivity_score: &mut u64,
+    validator_info: &ValidatorInfo,
+    state_ctxt: &StateContext,
+    spec: &ChainSpec,
+) -> Result<(), Error> {
+    if !validator_info.is_eligible {
+        return Ok(());
+    }
+
+    // Increase inactivity score of inactive validators
+    if validator_info.is_unslashed_participating_index(TIMELY_TARGET_FLAG_INDEX)? {
+        // Avoid mutating when the inactivity score is 0 and can't go any lower -- the common
+        // case.
+        if *inactivity_score == 0 {
+            return Ok(());
+        }
+        inactivity_score.safe_sub_assign(1)?;
+    } else {
+        inactivity_score.safe_add_assign(spec.inactivity_score_bias)?;
+    }
+
+    // Decrease the score of all validators for forgiveness when not during a leak
+    if !state_ctxt.is_in_inactivity_leak {
+        let deduction = min(spec.inactivity_score_recovery_rate, *inactivity_score);
+        inactivity_score.safe_sub_assign(deduction)?;
+    }
+
+    Ok(())
+}
+
+fn process_single_reward_and_penalty(
+    balance: &mut u64,
+    inactivity_score: &u64,
+    validator_info: &ValidatorInfo,
+    rewards_ctxt: &RewardsAndPenaltiesContext,
+    state_ctxt: &StateContext,
+    spec: &ChainSpec,
+) -> Result<(), Error> {
+    if !validator_info.is_eligible {
+        return Ok(());
+    }
+
+    let mut delta = Delta::default();
+    for flag_index in 0..NUM_FLAG_INDICES {
+        get_flag_index_delta(
+            &mut delta,
+            validator_info,
+            flag_index,
+            rewards_ctxt,
+            state_ctxt,
+        )?;
+    }
+    get_inactivity_penalty_delta(
+        &mut delta,
+        validator_info,
+        inactivity_score,
+        state_ctxt,
+        spec,
+    )?;
+
+    if delta.rewards != 0 || delta.penalties != 0 {
```

Comment:
Why is this if statement here? Does it make this function faster?

Reply:
It doesn't help on `stable`, but with `tree-states` we use a copy-on-write reference to the balance, and avoiding the mutation when it isn't necessary is a decent saving. See:

https://github.com/sigp/lighthouse/blob/928915c71885a2439f8fa3b3a22ef6b39e978339/consensus/state_processing/src/per_epoch_processing/single_pass.rs#L355-L359

That `make_mut` call is potentially expensive (it does a `VecMap` insert).